optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/manager/DefaultSolverManager.java

```diff
@@ -0,0 +1,188 @@
+            if (isProblemSubmitted(problemId)) {
+                throw new IllegalArgumentException("Problem (" + problemId + ") already exists.");
+            }
+            newSolverTask = new SolverTask<>(problemId, solverFactory.buildSolver(), planningProblem);
+            problemIdToSolverTaskMap.put(problemId, newSolverTask);
+            logger.info("A new solver task was created with problemId ({}).", problemId);
+        }
+
+        // TODO implement throttling
+        if (onBestSolutionChangedEvent != null) {
+            newSolverTask.addEventListener(
+                    bestSolutionChangedEvent ->
+                            eventHandlerExecutorService.submit(
+                                    () -> onBestSolutionChangedEvent.accept(bestSolutionChangedEvent.getNewBestSolution())));
+        }
+
+        CompletableFuture<Solution_> solverFuture = CompletableFuture.supplyAsync(newSolverTask::startSolving, solverExecutorService);
+        solverFuture.handle((solution_, throwable) -> {
+            if (throwable != null) {
+                logger.error("Exception while solving problem (" + problemId + ").", throwable.getCause());
+                if (onException != null) {
+                    eventHandlerExecutorService.submit(() -> onException.accept(throwable.getCause()));
+                }
+            }
+            return eventHandlerExecutorService.submit(() -> onSolvingEnded.accept(solution_));
+        });
+    }
+
+    @Override
+    public void stopSolver(Object problemId) {
+        logger.debug("Stopping solver of problemId ({}).", problemId);
+        SolverTask<Solution_> solverTask;
+        synchronized (this) {
+            if (!isProblemSubmitted(problemId)) {
+                throw new IllegalArgumentException("Problem (" + problemId + ") was not submitted.");
+            }
+            solverTask = problemIdToSolverTaskMap.get(problemId);
+        }
+        if (solverTask != null) {
+            solverTask.stopSolver();
+        }
+    }
+
+    @Override
+    public boolean isProblemSubmitted(Object problemId) {
+        return problemIdToSolverTaskMap.containsKey(problemId);
+    }
+
+    @Override
+    public Solution_ getBestSolution(Object problemId) {
+        logger.debug("Getting best solution of problemId ({}).", problemId);
+        SolverTask<Solution_> solverTask = problemIdToSolverTaskMap.get(problemId);
+        if (solverTask == null) {
+            logger.error("Problem (" + problemId + ") was not submitted.");
+            return null;
+        }
+        return solverTask.getBestSolution();
+    }
+
+    @Override
+    public Score<?> getBestScore(Object problemId) {
+        logger.debug("Getting best score of problemId ({}).", problemId);
+        SolverTask<Solution_> solverTask = problemIdToSolverTaskMap.get(problemId);
+        if (solverTask == null) {
+            logger.error("Problem (" + problemId + ") was not submitted.");
+            return null;
```

Comment:
Should throw an exception in my opinion. `SolverTask.getBestScore()` will return `null` if solution is uninitialized. So when you get `null` from `DefaultSolverManager.getBestScore()` how do you know your `problemId` is wrong or the solution is not yet initialized?

Reply:
@ge0ffrey and I discussed this point previously and decided to only log an error (I can't find where we documented the discussion and can't remember the arguments, but I'll bring it up again in our next meeting)